INSP: add lint inspection that annotates unnecessarily qualified paths

Kobzol opened the pull request:
Fixes: https://github.com/intellij-rust/intellij-rust/issues/5608

Related issue: https://github.com/intellij-rust/intellij-rust/issues/5888

changelog: Add inspection that finds unnecessarily qualified paths. It corresponds to the `unused_qualifications` lint.

vlad20012 commented on src/main/kotlin/org/rust/ide/inspections/lints/RsUnnecessaryQualificationsInspection.kt:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    private fun getUnnecessarilyQualifiedPath(path: RsPath): RsPath? {
```
Btw this looks a bit similar to my [trySimplifyPath](https://github.com/intellij-rust/intellij-rust/pull/7393/files#diff-6e12710c6434ae7f5debe4dd6682cf237be74e09d0680d12e891263e89bd9cfdR686) from #7393

Kobzol replied:
Indeed it does! It seems that your code also does some renaming? Which is probably not applicable here. I'm not sure if it's worth it to deduplicate these methods, as the intent is a bit different I guess.

dima74 commented on src/main/kotlin/org/rust/ide/inspections/fixes/RemovePathPrefixFix.kt:
```diff
@@ -0,0 +1,37 @@
+ *
+ * E.g. `foo::bar::baz` -> `baz`.
+ */
+class RemovePathPrefixFix(
```
Why not use `SubstituteTextFix.Companion#delete`?

Kobzol replied:
I wasn't aware that I can modify the fix name in such case. Thanks!

dima74 commented on src/main/kotlin/org/rust/ide/inspections/lints/RsUnnecessaryQualificationsInspection.kt:
```diff
@@ -0,0 +1,87 @@
+}
+
+private fun RsPath.canBeShortened(): Boolean {
+    return path != null || coloncolon != null
```
Why not `&&`?

Kobzol replied:
We also want to shorten paths like `::serde::Serialize`, so there can only be the double colon and no parent path.

dima74 commented on src/main/kotlin/org/rust/ide/inspections/lints/RsUnnecessaryQualificationsInspection.kt:
```diff
@@ -0,0 +1,87 @@
+                val target = getUnnecessarilyQualifiedPath(path)
+                if (target != null) {
+                    val unnecessaryPart = target.fullQualifier()
+                    val range = TextRange(0, unnecessaryPart.length)
```
This will not work when there are spaces or comments in the middle of the path (please add such test):
```rust
pub mod a {
    pub mod b {
        pub fn func() {}
    }
}

fn main() {
    use a::b::func;
    a :: b::func();
}
```

Consider using `target.referenceNameElement!!.startOffset` here instead

Kobzol replied:
Thanks, that's much easier indeed.

dima74 commented on src/main/kotlin/org/rust/ide/inspections/lints/RsUnnecessaryQualificationsInspection.kt:
```diff
@@ -0,0 +1,87 @@
+        var currentPath = path
+        val basePath = path.basePath()
+
+        while (true) {
```
I don't like a bit that we use cycle with mutable variables here. Maybe we can analyze not only root paths in visitor, so we will not need to do manual recursion here?

Kobzol replied:
But then the inspection could be triggered multiple times? For
`a::b::c` I'm walking the path from the right, and once I find some suffix (`c`, `b::c`, `a::b::c`) that can be resolved, I mark everything to the left as unnecessary. If I analyzed `a::b` and `a::b::c` separately, I could mark the `a::` path as unnecessary twice if both `b::c` and `c` were resolvable.

I'm also not sure if we could depend on the visitor on visiting the path in the right order. I can try to refactor the while loop to a more functional implementation though.

dima74 replied:
> But then the inspection could be triggered multiple times?

Yes, you are right

> I can try to refactor the while loop to a more functional implementation though.

Let's try. I would use something like this:
```kotlin
val qualifiers = generateSequence(path) { path.qualifier }
for (currentPath in qualifiers) {
    val currentName = currentPath.referenceName ?: break
    val pathText = "$currentName::${path.text.removePrefix(currentPath.text)}"
    ...
}
```

Kobzol commented:
I changed the implementation to be hopefully a bit readable (first time I have used `runningFold`! :D ). Although there are still some rough mainly around leading `::`. I also added a few more tests and rebased, since I needed changes from `master` to test the changes locally in `CLion`.

dima74 commented:
I tested the inspection on `cargo` and `rust-analyzer` projects, and found only one false-positive, good work 👍

Here is the false-positive:
```rust
fn main() {
    let _ = Option::<()>::None;
}
```

dima74 commented on src/main/kotlin/org/rust/ide/inspections/lints/RsUnnecessaryQualificationsInspection.kt:
```diff
@@ -0,0 +1,99 @@
+            val root = path.rootPath()
+            val useItem = path.parentOfType<RsUseItem>()
+
+            if (useItem == null && root == path && root.canBeShortened()) {
```
Let's inline root variable here, since it's a bit confusing to have two equal variables `root` and `path` inside `if`-block

Kobzol commented:
bors r=dima74

Undin commented:
bors retry
